synth/src/sampler.rs

```diff
@@ -1,101 +1,159 @@
 use anyhow::Result;
 use indicatif::{ProgressBar, ProgressStyle};
 use rand::SeedableRng;
-use serde_json::{Map, Value};
-use synth_core::Graph;
-use synth_core::{Name, Namespace};
+use synth_core::{Graph, Name, Namespace, Value};
 use synth_gen::prelude::*;
+use std::collections::{btree_map::Entry, BTreeMap};
+use std::convert::TryFrom;
 
-pub(crate) struct Sampler<'r> {
-    namespace: &'r Namespace,
+pub(crate) struct Sampler {
+    graph: Graph,
 }
 
-impl<'r> Sampler<'r> {
-    pub fn new(namespace: &'r Namespace) -> Self {
-        Self { namespace }
-    }
-
-    fn sampler_progress_bar(target: u64) -> ProgressBar {
-        let bar = ProgressBar::new(target as u64);
-        let style = ProgressStyle::default_bar()
-            .template("[{elapsed_precise}] {wide_bar} {pos}/{len} generated ({eta} remaining)");
-        bar.set_style(style);
-        bar
-    }
+fn sampler_progress_bar(target: u64) -> ProgressBar {
+    let bar = ProgressBar::new(target as u64);
+    let style = ProgressStyle::default_bar()
+        .template("[{elapsed_precise}] {wide_bar} {pos}/{len} generated ({eta} remaining)");
+    bar.set_style(style);
+    bar
+}
 
+impl Sampler {
     pub(crate) fn sample_seeded(
-        &self,
-        collection: Option<Name>,
+        self,
+        collection_name: Option<Name>,
         target: usize,
         seed: u64,
     ) -> Result<Value> {
-        let mut rng = rand::rngs::StdRng::seed_from_u64(seed);
+        let rng = rand::rngs::StdRng::seed_from_u64(seed);
+        let model = self.graph.aggregate();
+        let sample_strategy = SampleStrategy::new(collection_name, target);
+        sample_strategy.sample(model, rng)
+    }
+}
 
-        let graph = Graph::from_namespace(self.namespace)?;
-        let mut model = graph.into_iterator(&mut rng);
+impl TryFrom<&Namespace> for Sampler {
+    type Error = anyhow::Error;
+    fn try_from(namespace: &Namespace) -> Result<Self> {
+        Ok(Self {
+            graph: Graph::from_namespace(namespace)?,
+        })
+    }
+}
 
-        let target = target as u64;
-        let progress_bar = Self::sampler_progress_bar(target);
-        let mut output = Map::new();
+enum SampleStrategy {
+    Namespace(NamespaceSampleStrategy),
+    Collection(CollectionSampleStrategy),
+}
 
-        while progress_bar.position() < target {
-            let serializable = OwnedSerializable::new(&mut model);
-            let mut is_empty = true;
+impl SampleStrategy {
+    fn new(collection_name: Option<Name>, target: usize) -> Self {
+        match collection_name {
+            None => SampleStrategy::Namespace(NamespaceSampleStrategy {
+                target
+            }),
+            Some(name) => SampleStrategy::Collection(CollectionSampleStrategy {
+                name,
+                target,
+            })
+        }
+    }
 
-            let as_value = match (serde_json::to_value(serializable), model.restart()) {
```

Comment:
Will introduce a regression on the displaying of errors and their breaking down into generation-time errors whose cause is the user and generation-time errors whose cause is internal.

Reply:
Agreed. Let's fix this in a followup.